accepted/wasmtime-compile-time-builtins.md

```diff
@@ -0,0 +1,379 @@
+JavaScript execution environments.
+
+[js-string-builtins]: https://github.com/WebAssembly/js-string-builtins/blob/main/proposals/js-string-builtins/Overview.md
+
+# Motivation and Requirements
+[motivation]: #motivation
+
+We have two primary, and related, desiderata:
+
+1. **Remove call overheads for simple host functions.** For simple getter- and
+   setter-style functions, call overhead can dwarf the execution time of the
+   function body itself, particularly when we consider that the function
+   boundary hides information from Cranelift, inhibiting optimizations like GVN
+   and LICM between the caller and callee. Furthermore, with today's late-bound
+   imports, and without a tiering JIT compiler or mutating executable code,
+   direct calls to an imported function at the Wasm level must become indirect
+   calls through a [PLT/GOT]-like mechanism at the machine code level, imposing
+   additional overhead. We would like to eliminate these costs.
+
+2. **Enable performant, zero-copy manipulation of host buffers exposed to Wasm
+   as component model resources.** A host system often already has data in
+   external buffers, outside of a Wasm instance, and wants to grant Wasm access
+   to that data. This requires choosing between copying the data into the Wasm's
+   linear memory or giving the Wasm a file descriptor-style handle to an object
+   that represents that external buffer and its contents are read and written
+   via additional function imports. This latter approach successfully avoids the
+   cost of copying the data into Wasm linear memory, but introduces function
+   call overheads to all accesses. However, if the read and write accessor
+   functions are simple enough, we can inline them with this RFC's proposal,
+   eliminating the function call overheads while still avoiding copies.
+
+[PLT/GOT]: https://reverseengineering.stackexchange.com/a/1993
+
+Note that our goal is improving performance in the scenario where Wasm is
+importing host functions, and not the scenario where one Wasm module is linking
+to another Wasm module, satisfying its own imports with the other's
+exports. Host functions are compiled at a different moment in time and by a
+different compiler from Wasm compiled by Wasmtime. Host functions can access
+memory and resources outside of any Wasm instance's sandbox. These properties
+tie our hands and constrain our potential solutions.
+
+Additionally, note that the definition of compile-time builtins is pretty
+fundamentally `unsafe`: you are interacting with the guts of Wasm code
+generation and promising things like "I pinky swear that accessing this host
+memory at this pointer is equivalent to what my native host function would have
+otherwise done". Similarly, because these compile-time builtins must have the
+ability to load from and store to native memory, they cannot be portable across
+ISAs with different pointer widths.
```

Comment:
This seems to be in conflict with the later section describing intrinsics as always taking `u64` addresses (and ignoring the upper 32 bits on 32-bit targets). I like the latter a lot more and it seems nice to provide this portability boon, so perhaps this is just outdated text?

Reply:
Ah yes, sorry this is outdated text, will update momentarily.